pipelinedp4j/tests/com/google/privacy/differentialprivacy/pipelinedp4j/spark/SparkEncodersTest.kt

```diff
@@ -93,6 +94,16 @@ class SparkEncodersTest {
         assertThat(dataset.collectAsList()).containsExactlyElementsIn(input)
     }
 
+    @Test
+    fun tuple2sOf_tuple2sOf_isPossibleToCreateSparkCollectionOfThatType() {
```

Comment:
so this test passes? 

so before this change we had:

```
data.keyBy("KeyByPrivacyId", privacyIdEncoder) { it.privacyId() }
```

where data is `SparkCollection<ContributionWithPrivacyId<PrivacyIdT, PartitionKeyT>>` which is `SparkCollection<Pair<Pair<PrivacyIdT, PartitionKeyT>, Double>>`

The collection is created successfully, so Spark understands how to encode `Pair<Pair<PrivacyIdT, PartitionKeyT>, Double>` = `Pair<Pair<String, String>, Double>`

Then we do:

```
val valueEncoder = data.encoder()
        val keyEncoder = (outputType as SparkEncoder<K>).encoder
        val tupleEncoder = Encoders.tuple(keyEncoder, valueEncoder)
        val keyDataset = data.map(MapFunction { t: T ->  Tuple2(keyFn(t), t)}, tupleEncoder)
```

keyEncoder in our case is privacyIdEncoder, which is SparkEncoder<String>. Tuple encoder should be `TupleEncoder<String, Pair<Pair<PrivacyIdT, PartitionKeyT>, Double>>` = `TupleEncoder<String, Pair<Pair<String, String>, Double>>`. Seems a valid encoder that should work. And then we do `MapFunction { t: T ->  Tuple2(keyFn(t), t)}` which give an error: "java.lang.ClassCastException: class [B cannot be cast to class kotlin.Pair ([B is in module java.base of loader 'bootstrap'; kotlin.Pair is in unnamed module of loader 'app')". This error I don't understand fully. Do you undrstand why it doesn't work? I don't see a significant difference between Pair and any other type, including primitive types (for them tests pass).

Reply:
In previous code for `FrameworkCollection.keyBy()` we were returning `SparkTable<Tuple<K, V>>` and core api code is trying to cast into `FrameworkTable<Pair<K, V>>` somewhere, not sure where. I believe that is the cause of the problem and after changing it to `SparkTable<Pair<K, V>>` it works. I think this could be a problem even for BeamTable too. We can discuss about it over our chat.